[ffmpeg] fix linking with iconv feature

mcmtroffaes opened the pull request:
Linking applications against ffmpeg[iconv] currently fails with

```
libiconv.lib(iconv.c.obj) : error LNK2019: unresolved external symbol locale_charset referenced in function iconv_canonicalize
```

due to FindFFMPEG.cmake not adding libcharset.lib as a dependent library. This patch fixes that problem.

- What does your PR fix? See above.

- Which triplets are supported/not supported? Have you updated the CI baseline? Unchanged.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)? Yes.

mcmtroffaes committed:
[ffmpeg] fix linking with iconv feature

cenit commented:
Is there any interference between this iconv and the one searched on line 254 (it might be a problem only on macOS)?

Neumann-A commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -51,6 +51,9 @@ if(@ENABLE_ICONV@)
   get_target_property(ICONV_LIBRARY_RELEASE unofficial::iconv::libiconv IMPORTED_LOCATION_RELEASE)
   get_target_property(ICONV_LIBRARY_DEBUG unofficial::iconv::libiconv IMPORTED_LOCATION_DEBUG)
   list(APPEND FFMPEG_PLATFORM_DEPENDENT_LIBS "$<$<CONFIG:Debug>:${ICONV_LIBRARY_DEBUG}>$<$<CONFIG:Release>:${ICONV_LIBRARY_RELEASE}>")
+  get_target_property(CHARSET_LIBRARY_RELEASE unofficial::iconv::libcharset IMPORTED_LOCATION_RELEASE)
+  get_target_property(CHARSET_LIBRARY_DEBUG unofficial::iconv::libcharset IMPORTED_LOCATION_DEBUG)
+  list(APPEND FFMPEG_PLATFORM_DEPENDENT_LIBS "$<$<CONFIG:Debug>:${CHARSET_LIBRARY_DEBUG}>$<$<CONFIG:Release>:${CHARSET_LIBRARY_RELEASE}>")
```
`$<CONFIG:Debug>` and `$<$<CONFIG:Release>` is bad!
it should be `$<CONFIG:Debug>` and `$<NOT:$<CONFIG:Debug>>`

Please also use FindIconv.cmake (https://cmake.org/cmake/help/latest/module/FindIconv.html) instead of the unofficial target otherwise I might miss that change in #11776 which changes the build of Iconv to its native make based build system

mcmtroffaes replied:
Thanks for making me aware of that work. I can see if I can move to the new system to make it compatible with your patch.

I agree CONFIG:Debug & NOT:$<CONFIG:Debug> is more reliable: it covers RelWithDbgInfo and MinSizeRel too (ran into that myself on occasion). However, CONFIG:Debug/CONFIG:Release is used in a lot of places in vcpkg, including this portfile. Is there is an intention to change these throughout?

Neumann-A commented:
@cenit Iconv on macos should be part of the c library or not?

mcmtroffaes committed:
[ffmpeg] use official iconv cmake target (broken for now, needs fix in libiconv, see #11776)

mcmtroffaes committed:
[ffmpeg] fix typo in theora configuration

mcmtroffaes committed:
[ffmpeg] switch to $<NOT:$<CONFIG:Debug>> for selecting release libraries

mcmtroffaes commented:
The official iconv target appears broken (in particular, Iconv_LIBRARIES is invalid), so I cannot do the requested changes. I assume #11776 will fix this problem, so I'm putting this on hold until that work is merged.

mcmtroffaes commented:
Good news: it's working fine now. I thought something was off with the libraries but it was just cmake's unusual debug/optimized flag system for the charset libraries which confused me. Whilst at it I also noticed a typo in the theora config. I think I've addressed all issues now, and so this should be good to merge.

mcmtroffaes commented:
@cenit Judging from the code, it could be that Iconv can be removed from the macos specific section. However, I'm not 100% sure. At the moment, for macos, ffmpeg is *always* linked with the iconv libs even if the iconv feature is not requested, so perhaps it is required for macos. I don't have macos to test, unfortunately.

cenit commented:
Understood. I can try some tests next week. In the meantime thanks for your work. In any case, can I ask you to use find_dependency syntax also in the macOS section, since you have also revised some other problems of this cmake module?
I will tell you later if you can remove it or not anyway, but in the meantime at least there is a proper syntax there (better in handling REQUIRED keyword downstream, for example)

mcmtroffaes committed:
[ffmpeg] use find_dependency(Iconv) on macos to better handle downstream REQUIRED

mcmtroffaes commented:
Sure, I've added a commit to switch to use find_dependency instead. If you can test at some point whether it is necessary to specify the iconv dependency in that section (both static and dynamic builds, if possible), that would be swell.

cenit commented:
yes sure I hope to be able to get back to you on this topic very soon

mcmtroffaes committed:
Bump port version.

mcmtroffaes commented:
Rebased on latest master.

mcmtroffaes commented:
@cenit Any feedback from testing on mac, if you've had the chance already?

cenit commented:
@mcmtroffaes sorry for being so late

I tested your PR with the following added patch and it worked on my Mac. So you might be right, Iconv feature is not always active on osx anymore


```
--- a/ports/ffmpeg/FindFFMPEG.cmake.in
+++ b/ports/ffmpeg/FindFFMPEG.cmake.in
@@ -247,8 +247,7 @@ if(APPLE)
   if (NOT CV_UNIT)
       message(FATAL_ERROR "CoreVideo not found")
   endif()
-  find_dependency(Iconv)
-  list(APPEND FFMPEG_PLATFORM_DEPENDENT_LIBS ${VT_UNIT} ${AT_UNIT} ${SEC_UNIT} ${CF_UNIT} ${CM_UNIT} ${CV_UNIT} ${Iconv_LIBRARIES})
+  list(APPEND FFMPEG_PLATFORM_DEPENDENT_LIBS ${VT_UNIT} ${AT_UNIT} ${SEC_UNIT} ${CF_UNIT} ${CM_UNIT} ${CV_UNIT})
 endif()
 
 if(@ENABLE_AVCODEC@)
```

mcmtroffaes commented:
Great, many thanks for taking the time to test and for confirming what we thought to be the case. I'll push another patch to remove the Iconv dependency when it's not needed.

mcmtroffaes committed:
[ffmpeg] remove iconv dependency on mac when it's not needed

Patch originally provided by @cenit

mcmtroffaes commented:
@Neumann-A All comments from yourself and @cenit have been addressed now. Should be ready for final review & merging if no other issues turn up.

ras0219-msft commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -43,108 +43,106 @@ if(@ENABLE_BZIP2@)
-  get_target_property(ICONV_LIBRARY_RELEASE unofficial::iconv::libiconv IMPORTED_LOCATION_RELEASE)
-  get_target_property(ICONV_LIBRARY_DEBUG unofficial::iconv::libiconv IMPORTED_LOCATION_DEBUG)
-  list(APPEND FFMPEG_PLATFORM_DEPENDENT_LIBS "$<$<CONFIG:Debug>:${ICONV_LIBRARY_DEBUG}>$<$<CONFIG:Release>:${ICONV_LIBRARY_RELEASE}>")
+  find_dependency(Iconv)
```
Strictly speaking, `find_dependency()` isn't intended to be used from `FindXYZ` modules[1], however it should be fine in this case because it should never fail (yes, this is a preexisting bug). The fix is probably to make this `FFMPEG-config.cmake` and add a stub `FindFFMPEG` that just calls `find_package(FFMPEG CONFIG)` or somesuch.

I don't consider this to be a blocking issue for this PR.

[1] https://cmake.org/cmake/help/latest/module/CMakeFindDependencyMacro.html

ras0219-msft commented:
Thanks for the fixes!

mcmtroffaes commented:
Many thanks @ras0219-msft for merging, and pointing out the caveat around find_dependency. I missed that note in the documentation! I'll keep this in mind for the future. Yes, it would be easy to turn this into a ``FFMPEG-config.cmake`` style file eventually if that's needed.
